packages/taiko-client-rs/crates/protocol/src/shasta/manifest.rs

```diff
@@ -27,56 +30,187 @@ pub struct BlockManifest {
     #[serde(default)]
     pub prover_auth_bytes: Bytes,
-    /// Blocks bundled in this proposal.
+    /// Derivation sources included in this proposal.
     #[serde(default)]
-    pub blocks: Vec<BlockManifest>,
+    pub sources: Vec<DerivationSourceManifest>,
+}
+
+impl Default for ProposalManifest {
+    /// Create the default proposal manifest.
+    fn default() -> Self {
+        Self { prover_auth_bytes: Bytes::new(), sources: vec![DerivationSourceManifest::default()] }
+    }
 }
 
 impl ProposalManifest {
-    /// Encode the protocol proposal manifest following the Shasta protocol payload format.
+    /// Encode and compress the protocol proposal manifest following the Shasta protocol payload
+    /// format. Ref: https://github.com/taikoxyz/taiko-mono/blob/main/packages/protocol/docs/Derivation.md
+    pub fn encode_and_compress(&self) -> Result<Vec<u8>> {
+        encode_manifest_payload(self)
+    }
+
+    /// Decompress and decode a protocol proposal manifest from the Shasta protocol payload bytes.
     /// Ref: https://github.com/taikoxyz/taiko-mono/blob/main/packages/protocol/docs/Derivation.md
-    pub fn encode(&self) -> Result<Vec<u8>> {
-        let rlp_encoded = alloy_rlp::encode(self);
+    pub fn decompress_and_decode(bytes: &[u8], offset: usize) -> Result<Self> {
+        let Some(decoded) = decode_manifest_payload(bytes, offset)? else {
+            return Ok(ProposalManifest::default());
+        };
+
+        let mut decoded_slice = decoded.as_slice();
+        let manifest = <ProposalManifest as Decodable>::decode(&mut decoded_slice)
+            .map_err(|err| ProtocolError::Rlp(err.to_string()))?;
+
+        if manifest.sources.len() > PROPOSAL_MAX_BLOCKS {
+            return Ok(ProposalManifest::default());
+        }
+
+        Ok(manifest)
+    }
+}
+
+/// Encode a manifest into the Shasta protocol payload format.
+fn encode_manifest_payload<T>(manifest: &T) -> Result<Vec<u8>>
+where
+    T: Encodable,
+{
+    let rlp_encoded = alloy_rlp::encode(manifest);
+
+    let mut encoder = ZlibEncoder::new(Vec::new(), Compression::default());
+    encoder.write_all(&rlp_encoded)?;
+    let compressed = encoder.finish()?;
+
+    let mut output = Vec::with_capacity(64 + compressed.len());
+
+    let mut version_bytes = [0u8; 32];
+    version_bytes[31] = SHASTA_PAYLOAD_VERSION;
+    output.extend_from_slice(&version_bytes);
+
+    let len_bytes = U256::from(compressed.len()).to_be_bytes::<32>();
+    output.extend_from_slice(&len_bytes);
+    output.extend_from_slice(&compressed);
 
-        let mut encoder = ZlibEncoder::new(Vec::new(), Compression::default());
-        encoder.write_all(&rlp_encoded)?;
-        let compressed = encoder.finish()?;
+    Ok(output)
+}
 
-        let mut output = Vec::with_capacity(64 + compressed.len());
+/// Decode a manifest from the Shasta protocol payload format.
+fn decode_manifest_payload(bytes: &[u8], offset: usize) -> Result<Option<Vec<u8>>> {
```

Comment:
It doesn’t decode the blob the same way as the Go client. Is that expected?

Reply:
we are using: https://github.com/NethermindEth/Catalyst/blob/0c5d6627b5d6ff10f6be0ff7c3e1d56e479eab38/common/src/blob/blob_decoder.rs#L26

Reply:
To encode a blob in **`client-rs`** we use `SimpleCoder`:
https://github.com/taikoxyz/taiko-mono/blob/8ea6eb4554e9f7e0f5d87bc710a7f438ab114735/packages/taiko-client-rs/crates/proposer/src/transaction_builder.rs#L99

But in **`client-go`** we use "OptimismCoder" here:
https://github.com/taikoxyz/taiko-mono/blob/8ea6eb4554e9f7e0f5d87bc710a7f438ab114735/packages/taiko-client/proposer/transaction_builder/blob.go#L279

In `SplitToBlobs` we call `blob.FromData` which uses "OptimismCoder"

Reply:
Plz review this: https://github.com/taikoxyz/taiko-mono/pull/20368/commits/2160d2dd0e171fc8015fede440132bfdf4d271c3 , goal is to align with the Go impl.